collector/bgp.go

```diff
@@ -8,155 +8,88 @@ import (
 	bgpPeerDescs          = kingpin.Flag("collector.bgp.peer-descriptions", "Add the value of the desc key from the JSON formatted BGP peer description as a label to peer metrics. (default: disabled).").Default("False").Bool()
 	bgpPeerDescsText      = kingpin.Flag("collector.bgp.peer-descriptions.plain-text", "Use the full text field of the BGP peer description instead of the value of the JSON formatted desc key (default: disabled).").Default("False").Bool()
 	bgpAdvertisedPrefixes = kingpin.Flag("collector.bgp.advertised-prefixes", "Enables the frr_exporter_bgp_prefixes_advertised_count_total metric which exports the number of advertised prefixes to a BGP peer. This is an option for older versions of FRR that don't have PfxSent field (default: disabled).").Default("False").Bool()
 )
 
-// BGPCollector collects BGP metrics, implemented as per prometheus.Collector interface.
-type BGPCollector struct{}
-
-// NewBGPCollector returns a BGPCollector struct.
-func NewBGPCollector() *BGPCollector {
-	return &BGPCollector{}
+func init() {
+	registerCollector(bgpSubsystem, enabledByDefault, NewBGPCollector)
+	registerCollector(bgpSubsystem+"6", enabledByDefault, NewBGP6Collector)
+	registerCollector(bgpSubsystem+"l2vpn", enabledByDefault, NewBGPL2VPNCollector)
 }
 
-// Name of the collector. Used to populate flag name.
-func (*BGPCollector) Name() string {
-	return bgpSubsystem
-}
-
-// Help describes the metrics this collector scrapes. Used to populate flag help.
-func (*BGPCollector) Help() string {
-	return "Collect BGP Metrics"
-}
-
-// EnabledByDefault describes whether this collector is enabled by default. Used to populate flag default.
-func (*BGPCollector) EnabledByDefault() bool {
-	return true
-}
-
-// Describe implemented as per the prometheus.Collector interface.
-func (*BGPCollector) Describe(ch chan<- *prometheus.Desc) {
-	for _, desc := range getBgpDesc() {
-		ch <- desc
-	}
+type bgpCollector struct {
+	logger log.Logger
 }
 
-// Collect implemented as per the prometheus.Collector interface.
-func (c *BGPCollector) Collect(ch chan<- prometheus.Metric) {
-	bgpErrors = []error{}
-	collectBGP(ch, "ipv4")
+// NewBGPCollector collects BGP metrics, implemented as per the Collector interface.
+func NewBGPCollector(logger log.Logger) (Collector, error) {
+	return &bgpCollector{logger: logger}, nil
 }
 
-// CollectErrors returns what errors have been gathered.
-func (*BGPCollector) CollectErrors() []error {
-	return bgpErrors
+// Update implemented as per the Collector interface.
+func (c *bgpCollector) Update(ch chan<- prometheus.Metric) error {
+	return collectBGP(ch, "ipv4", c.logger)
 }
 
-// CollectTotalErrors returns total errors.
-func (*BGPCollector) CollectTotalErrors() float64 {
-	return totalBGPErrors
+type bgp6Collector struct {
```

Comment:
You can merge the `bgp6Collector` and `bgpCollector` structs by adding a variable `AFI string` to it. Then you can
* drop the `Update` function for `bgp6Collector`
* move collectBGP into the `Update` function

Reply:
Nice! Done.